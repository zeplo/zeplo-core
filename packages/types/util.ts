
export type Maybe<T> = T | null

export type JsonValue =  boolean | number | string | null | JsonArray | JsonMap
export interface JsonMap { [key: string]: JsonValue }
export interface JsonArray extends Array<JsonValue> {}

export type DeepPartial<T> = T extends Function ? T : (T extends object ? { [P in keyof T]?: DeepPartial<T[P]>; } : T);

/**
 * Get all of the keys to which U can be assigned.
 */
export type OnlyKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? K : never
}[keyof T];

/**
 * Get the interface containing only properties to which U can be assigned.
 */
export type OnlyUndefined<T> = {
  [K in OnlyKeys<T, undefined>]: T[K]
}

/**
 * Get all of the keys except those to which U can be assigned.
 */
export type ExcludeKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? never : K
}[keyof T];

/**
 * Get the interface containing no properties to which U can be assigned.
 */
export type ExcludeUndefined<T> = {
  [K in ExcludeKeys<T, undefined>]: T[K]
}

/**
 * Get the interface where all properties are optional.
 */
export type Optional<T> = {[K in keyof T]?: T[K]};

/**
 * Get the interface where properties that can be assigned undefined are
 * also optional.
 */
export type UndefinedOptional<T> = ExcludeUndefined<T> & Optional<OnlyUndefined<T>>

export type DeepUndefinedOptional<T> = T extends (...args: any[]) => any
  ? T
  : T extends any[]
  ? _DeepRequiredArray<T[number]>
  : T extends object
  ? _DeepRequiredObject<UndefinedOptional<T>>
  : T;

 export interface _DeepRequiredArray<T>
  extends Array<DeepUndefinedOptional<T>> {}

  export type _DeepRequiredObject<T> = {
  [P in keyof T]: DeepUndefinedOptional<T[P]>
}