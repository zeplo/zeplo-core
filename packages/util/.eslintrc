{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parser":  "@typescript-eslint/parser",
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
    "mocha": true,
    "jest": true
  },
  "rules": {
    "semi": ["error", "never"],
    "no-underscore-dangle": 0,
    "space-before-function-paren": ["error", "always"],
    "arrow-body-style": 0,
    "no-use-before-define": ["error", { "functions": false }],
    "arrow-parens": 0,
  }
}
