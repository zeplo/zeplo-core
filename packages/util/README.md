# @zeplo/util

Utility functions that are used across packages.