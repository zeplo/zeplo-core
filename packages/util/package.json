{
  "name": "@zeplo/util",
  "version": "2.1.0",
  "main": "dist/index.js",
  "files": [
    "dist/*"
  ],
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "env-cmd ts-node",
    "build": "tsc",
    "prepublish": "yarn clean && yarn build"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/fs-extra": "^9.0.13",
    "@types/lodash": "^4.14.178",
    "@types/node": "^17.0.18",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "env-cmd": "^10.1.0",
    "eslint": "^8.9.0",
    "rimraf": "^3.0.2",
    "ts-node": "^10.5.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@zeplo/errors": "^2.0.8",
    "@zeplo/types": "^2.1.0",
    "express": "^4.17.3",
    "express-async-handler": "^1.2.0",
    "lodash": "^4.17.21",
    "uuid": "^8.3.2"
  }
}
