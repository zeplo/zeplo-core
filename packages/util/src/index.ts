export { cleanQuery, cleanHeaders } from './clean'
export { 
  parseRequest, 
  parseBody, 
  getBackoff, 
  getStartTimeFromFeatures,
  getRawFeatures,
  parseRawRequestObject, 
  createRequestJob ,
  getValueFromQueryOrHeader,
} from './parse'
export { 
  createStepsJobs,
  hasCyclicalDependency,
} from './steps'
export { 
  requestHandler,
  jsonRequestHandler,
  queryToString,
  queryToNumber,
  parseQuery,
} from './request'