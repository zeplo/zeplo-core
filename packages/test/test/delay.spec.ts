/**
 * @jest-environment node
 */

import { omit } from 'lodash'
import { createWorkspace } from '#/util/data'
import waitFor from '#/util/waitFor'
import wait from '#/util/wait'
import queue from '#/util/queue'

jest.setTimeout(60000)

describe('delay.spec', () => {
  const workspace = `test-workspace-delay-${Date.now()}`
  const token = `test-token-delay-${Date.now()}`

  let cleanup: () => Promise<void>
  beforeAll(async () => {
    cleanup = await createWorkspace(workspace, token)
  })
  afterAll(() => cleanup())

  test('add delayed request using _delay param', async () => {
    const before = Date.now() / 1000
    const delay = 10
    const res = await queue.get(`/https://postman-echo.com/get?_delay=${delay}&_token=${token}`)
    const after = Date.now() / 1000

    expect(res.data).toHaveProperty('id')

    const pending = await queue.get(`/requests/${res.data.id}?_token=${token}`)
    expect(pending.data).toMatchObject({
      status: 'PENDING',
    })
    expect(pending.data.start).toBeGreaterThan(before + delay)
    expect(pending.data.start).toBeLessThan(after + delay + 1)

    // Can take 5 seconds for log to flush!
    // await wait(6000)

    // const pending2 = await queue.get(`/requests?_token=${token}`)
    // expect(pending2.data[0]).toMatchObject({
    //   status: 'PENDING',
    // })

    await wait(10000)

    const req = await waitFor(
      () => queue.get(`/requests/${res.data.id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'GET' },
      response: { headers: { connection: 'close' }, status: 200 },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('add delayed request using X-Zeplo-Delay header', async () => {
    const before = Date.now() / 1000
    const delay = 10
    const res = await queue.get(`/https://postman-echo.com/get?_token=${token}`, {
      headers: { 'X-Zeplo-Delay': delay },
    })
    const after = Date.now() / 1000

    expect(res.data).toHaveProperty('id')

    const pending = await queue.get(`/requests/${res.data.id}?_token=${token}`)
    expect(pending.data).toMatchObject({
      status: 'PENDING',
    })
    expect(pending.data.start).toBeGreaterThan(before + delay - 1)
    expect(pending.data.start).toBeLessThan(after + delay + 1)

    await wait(delay * 1000)

    const req = await waitFor(
      () => queue.get(`/requests/${res.data.id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'GET' },
      response: { headers: { connection: 'close' }, status: 200 },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })
})
