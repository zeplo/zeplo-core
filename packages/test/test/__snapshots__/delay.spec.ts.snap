// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delay.spec add delayed request using _delay param 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;

exports[`delay.spec add delayed request using X-Zeplo-Delay header 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;
